### PR TITLE
[mellanox] Facilitate automatic integration of new hw-mgmt

### PR DESCRIPTION
**Pull request by vivekrnv**
#### Why I did it

Facilitate Automatic integration of new hw-mgmt version into SONiC. 

**Inputs to the Script:**
1) `MLNX_HW_MANAGEMENT_VERSION` Eg: `7.0040.5202`
2) `CREATE_BRANCH: (y|n)` Creates a branch instead of a commit (optional, default: n) 
3) `BRANCH_SONIC`:  Only relevant when CREATE_BRANCH is y. `Default: master`. 

Note: These should be provided through `SONIC_OVERRIDE_BUILD_VARS ` parameter

**Output:**
1) Script creates a commit (in each of sonic-buildimage, sonic-linux-kernel) with all the changes required for upgrading the hw-management version to a version provided by  `MLNX_HW_MANAGEMENT_VERSION`

**Brief Summary of the changes made:**
1) `MLNX_HW_MANAGEMENT_VERSION` flag in the hw-management.mk file
2)  hw-mgmt submodule is updated to the corresponding version
3)  Updates are made to non-upstream-patches/patches and series.patch file
4)  series, kconfig-inclusion and kconfig-exclusion files can be updated in the sonic-linux-kernel repo
5)  sonic-linux-kernel/patches folder is updated with the corresponding upstream patches
6)  Based on the inputs, there could be a branch seen in the local for each of the repo's. Branch is named as `<branch>_<parent_commit>_integrate_<hw_mgmt_version>`

#### How I did it

1) Added a new make target which can be invoked by calling `make integrate-mlnx-hw-mgmt`

```
user@server:/sonic-buildimage$ git rev-parse --abbrev-ref HEAD
master_23193446a_integrate_7.0020.5052
user@server:/sonic-buildimage$ git log --oneline -n 2
f66e01867 (HEAD -> master_23193446a_integrate_V.7.0020.5052, show) Intgerate HW-MGMT V.7.0020.5052 Changes
23193446a (master_intg_hw_mgmt) Update logic

user@server:/sonic-buildimage/src/sonic-linux-kernel$ git rev-parse --abbrev-ref HEAD
master_6847319_integrate_7.0020.4104
user@server:/sonic-buildimage/src/sonic-linux-kernel$ git log --oneline -n 2
6094f71 (HEAD -> master_6847319_integrate_V.7.0020.5052) Intgerate HW-MGMT V.7.0020.5052 Changes
6847319 (origin/master, origin/HEAD) Read ID register for optoe1 to find pageable bit in optoe driver  (#308)
```

Changes made will be summarized under `sonic-buildimage/integrate-mlnx-hw-mgmt_user.out` file. Debugging and troubleshooting output is written to `sonic-buildimage/integrate-mlnx-hw-mgmt.log` files

Jenkins output file: 
http://nbu-nfs.mellanox.com/auto/mtrsysgwork/vkarri/integrate-mlnx-hw-mgmt_user.out

stdout for debugging: 
http://nbu-nfs.mellanox.com/auto/mtrsysgwork/vkarri/integrate-mlnx-hw-mgmt.log

#### Limitations:
1) Assumes the changes would only work for amd64
2) Assumes the non-upstream patches in mellanox only belong to hw-mgmt

#### How to verify it

Build the Kernel

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111
- [ ] 202205
- [ ] 202211

### CODE REVIEW

**Review by dgsudharsan — CHANGES_REQUESTED** [review]

File: platform/mellanox/hw-management.mk [hunk 1]
```diff
@@ -24,3 +24,84 @@ MLNX_HW_MANAGEMENT = hw-management_1.mlnx.$(MLNX_HW_MANAGEMENT_VERSION)_$(CONFIG
...
 $(MLNX_HW_MANAGEMENT)_DEPENDS += $(LINUX_HEADERS) $(LINUX_HEADERS_COMMON)
 SONIC_MAKE_DEBS += $(MLNX_HW_MANAGEMENT)
+
+TEMP_HW_MGMT_DIR = /tmp/hw_mgmt
+PTCH_DIR = $(TEMP_HW_MGMT_DIR)/patch_dir/
+NON_UP_PTCH_DIR = $(TEMP_HW_MGMT_DIR)/non_up_patch_dir/
+PTCH_LIST  = $(TEMP_HW_MGMT_DIR)/series
+KCFG_LIST = $(TEMP_HW_MGMT_DIR)/kconfig
+HWMGMT_NONUP_LIST = $(BUILD_WORKDIR)/$($(MLNX_HW_MANAGEMENT)_SRC_PATH)/hwmgmt_nonup_patches
+JENKINS_OUTFILE = $(BUILD_WORKDIR)/integrate-mlnx-hw-mgmt_jenkins.out
```
[the remark this answers is not in the document]
Comment (vivekrnv) [reply platform/mellanox/hw-management.mk]: Updated

**Review by dgsudharsan — CHANGES_REQUESTED** [review]

File: platform/mellanox/integration-scripts/hwmgmt_kernel_patches.py [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+
+    def write_series_diff(self):
+        diff = difflib.unified_diff(Data.final_slk_series, Data.non_up_slk_series, fromfile='a/patch/series', tofile="b/patch/series", lineterm="\n")
+        lines = []
+        for line in diff:
+            lines.append(line)
+        print("\n -> POST: final series.diff \n{}".format("".join(lines)))
+        FileHandler.write_lines(os.path.join(self.args.build_root, NON_UP_PATCH_DIFF), lines, True)
+    
+    def find_conflicts(self):
```
Comment (dgsudharsan) [inline platform/mellanox/integration-scripts/hwmgmt_kernel_patches.py:224]: Can we rename to check_kconfig_conflicts?

File: platform/mellanox/integration-scripts/hwmgmt_kernel_patches.py [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+                    has_conflict = True
+        return has_conflict
+
+    def handle_exclusions(self):
+        new_lines = []
+        curr_hdr = ""
+        for line_raw in Data.kcfg_exclude:
+            line = line_raw.strip()
+            should_exclude = False
+            if line_raw.strip():
```
Comment (dgsudharsan) [inline platform/mellanox/integration-scripts/hwmgmt_kernel_patches.py:256]: if line: ?

File: platform/mellanox/integration-scripts/hwmgmt_kernel_patches.py [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+                match = re.search(KCFG_HDR_RE, line)
+                if match:
+                    curr_hdr = match.group(1)
+                else:
+                    for (kcfg, _) in Data.updated_kcfg:
+                        if kcfg == line and curr_hdr in HDRS:
```
Comment (dgsudharsan) [inline platform/mellanox/integration-scripts/hwmgmt_kernel_patches.py:262]: The current design has limitation that hw-mgmt provided configs belong to amd64 and common. Hence the script needs to be revisited when support for arm64 is introduced

File: platform/mellanox/hw-management.mk [hunk 5]
```diff
@@ -24,3 +24,117 @@ MLNX_HW_MANAGEMENT = hw-management_1.mlnx.$(MLNX_HW_MANAGEMENT_VERSION)_$(CONFIG
...
+	echo -en '\n###-> hw-mgmt submodule update <-###\n' >> ${USER_OUTFILE}
+	git diff --no-color  HEAD~1 HEAD -- $($(MLNX_HW_MANAGEMENT)_SRC_PATH)/hw-mgmt >> ${USER_OUTFILE}
+
+	echo -en '\n###-> hw-management make file version change <-###\n' >> ${USER_OUTFILE}
+	git diff --no-color  HEAD~1 HEAD -- $(PLATFORM_PATH)/hw-management.mk >> ${USER_OUTFILE}
+	
+	echo -en '\n###-> Summary of buildimage changes <-###\n' >> ${USER_OUTFILE}
+	git diff --no-color  HEAD~1 HEAD --stat --output=${TMPFILE_OUT} -- $(PLATFORM_PATH)
+	cat ${TMPFILE_OUT} | tee -a ${USER_OUTFILE}
+	popd
```
Comment (dgsudharsan) [inline platform/mellanox/hw-management.mk:136]: Should we do git push if there is CREATE_BRANCH?

File: platform/mellanox/integration-scripts/hwmgmt_kernel_patches.py [hunk 6]
```diff
@@ -0,0 +1,314 @@
+#!/usr/bin/python
```
Comment (dgsudharsan) [inline platform/mellanox/integration-scripts/hwmgmt_kernel_patches.py:1]: Can we update to /usr/bin/env python?

File: platform/mellanox/integration-scripts/tests/data/expected_data/series.patch
Comment (dgsudharsan) [inline platform/mellanox/integration-scripts/tests/data/expected_data/series.patch:1]: Why is this empty file needed?

File: platform/mellanox/integration-scripts/tests/test_filehandler.py [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+    def test_write_lines_marker(self, mock_read_raw, mock_write_lines_marker):
+        global LINES_READ
+        global LINES_WRITE
+        LINES_READ = MOCK_SLK_KCFG.splitlines(True)
+        LINES_WRITE = FINAL_MOCK_SLK_KCFG.splitlines(True)
+
+        list_opts = KCFG.parse_opts_strs(UPDATED_MLNX_KCFG.split("\n"))
+        writable_opts = KCFG.get_writable_opts(list_opts)
+
+        FileHandler.write_lines_marker("", writable_opts, marker="mellanox")
```
Comment (dgsudharsan) [inline platform/mellanox/integration-scripts/tests/test_filehandler.py:146]: Don't see any asserts or checks in this test. Can you please add?

File: platform/mellanox/integration-scripts/tests/test_filehandler.py [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+    def test_read_kconfig_inclusion(self, mock_read_raw, mock_write_lines_marker):
+        global LINES_READ
+        LINES_READ = FINAL_MOCK_SLK_KCFG.splitlines(True)
+        opts = FileHandler.read_kconfig_inclusion("")
+
+        global LINES_WRITE
+        LINES_WRITE = UPDATED_MLNX_KCFG.splitlines()
+        writable_opts = KCFG.get_writable_opts(KCFG.parse_opts_strs(opts))
+        print(writable_opts)
+        FileHandler.write_lines("", writable_opts)
```
Comment (dgsudharsan) [inline platform/mellanox/integration-scripts/tests/test_filehandler.py:159]: Don't see any asserts or checks in this test. Can you please add?

**Review by dgsudharsan — CHANGES_REQUESTED** [review]

File: platform/mellanox/hw-management.mk (hunk 5 above)
Comment (vivekrnv) [reply platform/mellanox/hw-management.mk:136]: We don't know the remote to push. User should do this manually as of now

File: platform/mellanox/integration-scripts/tests/test_filehandler.py (hunk 8 above)
Comment (vivekrnv) [reply platform/mellanox/integration-scripts/tests/test_filehandler.py:159]: asserts or checks are present in writer_mock & reader_mock

File: platform/mellanox/integration-scripts/tests/test_filehandler.py (hunk 7 above)
Comment (vivekrnv) [reply platform/mellanox/integration-scripts/tests/test_filehandler.py:146]: asserts or checks are present in writer_mock & reader_mock

File: platform/mellanox/integration-scripts/hwmgmt_kernel_patches.py [hunk 9]
```diff
@@ -0,0 +1,314 @@
...
+        for patch in Data.old_down + Data.new_up:
+            file_n = os.path.join(self.args.build_root, os.path.join(NON_UP_PATCH_LOC, patch))
+            if os.path.isfile(file_n):
+                print(patch)
+                os.remove(file_n)
+
+        # Make sure the dir is now empty as all these patches are of hw-mgmt
+        files = FileHandler.read_dir(os.path.join(self.args.build_root, NON_UP_PATCH_LOC), "*.patch")
+        if files:
+            print("\n -> POST: Patches Remaining in {}, likely SDK non-upstream patches \n{}".format(NON_UP_PATCH_LOC, files))
```
[the remark this answers is not in the document]
Comment (vivekrnv) [reply platform/mellanox/integration-scripts/hwmgmt_kernel_patches.py:188]: Actually, will fail if other patches exist. Current the logic assumes the patches in non-up directory only belong to hw-mgmt. Will revisit when the SDK might have non-upstream patches

File: platform/mellanox/integration-scripts/hwmgmt_kernel_patches.py [hunk 10]
```diff
@@ -0,0 +1,314 @@
...
+        print("\n -> POST: final series.diff \n{}".format("".join(lines)))
+        FileHandler.write_lines(os.path.join(self.args.build_root, NON_UP_PATCH_DIFF), lines, True)
+    
+    def find_conflicts(self):
+        # current config under mellanox marker
+        old_mlnx_kcfg =  FileHandler.read_kconfig_inclusion(os.path.join(self.args.build_root, SLK_KCONFIG))
+        old_mlnx_kcfg = KCFG.parse_opts_strs(old_mlnx_kcfg)
+
+        print("-> INFO: [common] + [amd64] Kconfig: \n{}".format("\n".join(KCFG.get_writable_opts(Data.current_kcfg))))
+        print("-> INFO: current mellanox marker Kconfig: \n{}".format("\n".join(KCFG.get_writable_opts(old_mlnx_kcfg))))
```
[the remark this answers is not in the document]
Comment (vivekrnv) [reply platform/mellanox/integration-scripts/hwmgmt_kernel_patches.py:230]: Current would mean old, new is phrased as updated

File: platform/mellanox/integration-scripts/hwmgmt_kernel_patches.py (hunk 4 above)
Comment (vivekrnv) [reply platform/mellanox/integration-scripts/hwmgmt_kernel_patches.py:262]: Yes, have to check with hw-mgmt team.

File: platform/mellanox/integration-scripts/tests/data/expected_data/series.patch
Comment (vivekrnv) [reply platform/mellanox/integration-scripts/tests/data/expected_data/series.patch:1]: updated

**Comment by vivekrnv** [comment]
Handled all the comments, updated description to document limitations and update makefile to only commit when there are changes

**Review by dgsudharsan — CHANGES_REQUESTED** [review]
Please address the remaining unresolved comments

File: platform/mellanox/integration-scripts/hwmgmt_kernel_patches.py [hunk 11]
```diff
@@ -0,0 +1,314 @@
...
+        Data.old_down = FileHandler.read_strip_minimal(self.args.current_non_up_patches)
+
+        # Read the new kcfg
+        new_cfg = FileHandler.read_kconfig_inclusion(self.args.config_inclusion, None)
+        Data.updated_kcfg = KCFG.parse_opts_strs(new_cfg)
+
+        # entire current config, [common] + [amd64]
+        all_kcfg = FileHandler.read_kconfig_parser(os.path.join(self.args.build_root, SLK_KCONFIG))
+        Data.current_kcfg = []
+        for hdr in HDRS:
```
[the remark this answers is not in the document]
Comment (vivekrnv) [reply platform/mellanox/integration-scripts/hwmgmt_kernel_patches.py:128]: arm64 support will be an enhancement

File: platform/mellanox/integration-scripts/hwmgmt_kernel_patches.py [hunk 12]
```diff
@@ -0,0 +1,314 @@
...
+        
+        if not action.check():
+            print("-> ERR: Argument Checks Failed")
+            sys.exit(1)
+
+        return action
+
+    def check(self):
+        if not self.args.config_inclusion:
+            print("-> ERR: onfig_inclusion is missing")
```
[the remark this answers is not in the document]
Comment (vivekrnv) [reply platform/mellanox/integration-scripts/hwmgmt_kernel_patches.py]: updated

File: platform/mellanox/integration-scripts/helper.py [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+        if not opt.startswith("CONFIG"):
+            print("-> DEBUG: Malformed kconfig opt, {}".format(opt))
+            return ()
+
+        tmp = opt.split("=")
+        if len(tmp) != 2:
+            print("-> DEBUG: Malformed kconfig opt, {}".format(opt))
+            return ()
+        
+        if not tmp[0].startswith("CONFIG"):
```
[the remark this answers is not in the document]
Comment (vivekrnv) [reply platform/mellanox/integration-scripts/helper.py]: Yes, removed the check

**Comment by vivekrnv** [comment]
> Please address the remaining unresolved comments

Addressed

**Review by dgsudharsan — APPROVED** [review]

**Pull request closed without merging**